libs/methodologies/bold/rule-processors/mass-id/src/certificate-uniqueness-check/certificate-uniqueness-check.processor.ts

```diff
@@ -0,0 +1,194 @@
+  massIdCertificateDocuments: Document[];
+  relatedMassIdAuditDocuments: Document[];
+}
+
+export const RESULT_COMMENTS = {
+  APPROVED: `The ${MASS_ID} is not linked to a valid ${MASS_ID} Certificate`,
+} as const;
+
+export class CertificateUniquenessCheck extends RuleDataProcessor {
+  private readonly massIdCertificateMatcher: DocumentMatcher;
+
+  private readonly methodologyName: BoldMethodologyName;
+
+  readonly errorProcessor = new CertificateUniquenessCheckProcessorErrors();
+
+  constructor(
+    massIdCertificateMatcher: DocumentMatcher,
+    methodologyName: BoldMethodologyName,
+  ) {
+    super();
+
+    this.massIdCertificateMatcher = massIdCertificateMatcher;
+    this.methodologyName = methodologyName;
+  }
+
+  private evaluateResult({
+    creditDocuments,
+    massIdCertificateDocuments,
+    relatedMassIdAuditDocuments,
+  }: RuleSubject): EvaluateResultOutput {
+    if (this.hasSomeValidCreditDocument(creditDocuments)) {
+      throw this.errorProcessor.getKnownError(
+        this.errorProcessor.ERROR_MESSAGE
+          .MASS_ID_DOCUMENT_HAS_A_VALID_CREDIT_DOCUMENT,
+      );
+    }
+
+    if (this.hasSomeValidCertificateDocument(massIdCertificateDocuments)) {
+      throw this.errorProcessor.getKnownError(
+        this.errorProcessor.ERROR_MESSAGE.MASS_ID_DOCUMENT_HAS_A_VALID_CERTIFICATE_DOCUMENT(
+          assert<NonEmptyString>(this.massIdCertificateMatcher.match.type),
+        ),
+      );
+    }
+
+    if (this.hasSomeApprovedMassIdAuditDocument(relatedMassIdAuditDocuments)) {
+      throw this.errorProcessor.getKnownError(
+        this.errorProcessor.ERROR_MESSAGE.MASS_ID_DOCUMENT_HAS_A_AUDIT_FOR_SAME_METHODOLOGY_NAME(
+          this.methodologyName,
+        ),
+      );
+    }
+
+    return {
+      resultComment: RESULT_COMMENTS.APPROVED,
+      resultStatus: RuleOutputStatus.APPROVED,
+    };
+  }
+
+  private async getRuleSubject(
+    documentQuery: DocumentQuery<Document> | undefined,
+    ruleInput: RuleInput,
+  ): Promise<RuleSubject> {
+    const creditDocuments: Document[] = [];
+    const massIdCertificateDocuments: Document[] = [];
+    let relatedMassIdAuditDocuments: Document[] = [];
+
+    await documentQuery?.iterator().each(({ document }) => {
+      const documentReference = mapDocumentReference(document);
+
+      if (CREDITS.matches(documentReference)) {
+        creditDocuments.push(document);
+      }
+
+      if (this.massIdCertificateMatcher.matches(documentReference)) {
+        massIdCertificateDocuments.push(document);
+      }
+
+      if (MASS_ID_AUDIT.matches(documentReference)) {
+        relatedMassIdAuditDocuments.push(document);
+      }
+    });
+
+    relatedMassIdAuditDocuments = relatedMassIdAuditDocuments.filter(
+      (document) => document.id !== ruleInput.documentId,
+    );
+
+    return {
+      creditDocuments,
+      massIdCertificateDocuments,
+      relatedMassIdAuditDocuments,
+    };
+  }
+
+  private hasSomeApprovedMassIdAuditDocument(
```

Comment:
I think we should have a clearer name
```suggestion
  private hasApprovedMassIdAuditForTheSameMethodology(
```

Reply:
https://github.com/carrot-foundation/methodology-rules/pull/225/commits/2456eaa90e0c43fe467f7c1fb9ffa6a9f335dce6